ros2_batch_job/__main__.py

```diff
@@ -237,58 +254,65 @@ def process_coverage(args, job):
-    info("ament.py build returned: '{0}'".format(ret_build))
+    print('# BEGIN SUBSECTION: colcon build')
+    cmd = [
+        args.colcon_script, 'build',
+        '--base-paths', '"%s"' % args.sourcespace,
+        '--build-base', '"%s"' % args.buildspace,
+        '--install-base', '"%s"' % args.installspace,
+    ] + (['--merge-install'] if not args.isolated else []) + \
+        args.build_args
+
+    cmake_args = ['" -DBUILD_TESTING=1"']
+    if args.cmake_build_type:
+        cmake_args.append(
+            '" -DCMAKE_BUILD_TYPE=' + args.cmake_build_type + '"')
+    if '--cmake-args' in cmd:
+        index = cmd.index('--cmake-args')
+        cmd[index + 1:index + 1] = cmake_args
+    else:
+        cmd.append('--cmake-args')
+        cmd.extend(cmake_args)
+
+    if coverage:
+        ament_cmake_args = [
+            '" -DCMAKE_CXX_FLAGS=${CMAKE_CXX_FLAGS} ' + gcov_flags + '"',
+            '" -DCMAKE_C_FLAGS=${CMAKE_C_FLAGS} ' + gcov_flags + '"']
+        if '--ament-cmake-args' in cmd:
+            index = cmd.index('--ament-cmake-args')
+            cmd[index + 1:index + 1] = ament_cmake_args
+        else:
+            cmd.append('--ament-cmake-args')
+            cmd.extend(ament_cmake_args)
+
+    ret_build = job.run(cmd, shell=True)
+    info("colcon build returned: '{0}'".format(ret_build))
     print('# END SUBSECTION')
     if ret_build:
         return ret_build
 
-    print('# BEGIN SUBSECTION: ament test')
-    # Run tests
+    print('# BEGIN SUBSECTION: colcon test')
     ret_test = job.run([
-        '"%s"' % job.python, '-u', '"%s"' % ament_py, 'test',
-        '"%s"' % args.sourcespace,
-        '--build-space', '"%s"' % args.buildspace,
-        '--install-space', '"%s"' % args.installspace,
-        # Skip building and installing, since we just did that successfully.
-        '--skip-build', '--skip-install',
-        # Ignore return codes that indicate test failures;
-        # they'll be picked up later in test reporting.
-        '--ignore-return-codes',
-    ] + (['--isolated'] if args.isolated else []) + args.ament_test_args,
+        args.colcon_script, 'test',
+        '--base-paths', '"%s"' % args.sourcespace,
+        '--build-base', '"%s"' % args.buildspace,
+        '--install-base', '"%s"' % args.installspace,
+    ] + (['--merge-install'] if not args.isolated else []) + args.test_args,
         exit_on_error=False, shell=True)
-    info("ament.py test returned: '{0}'".format(ret_test))
+    info("colcon test returned: '{0}'".format(ret_test))
     print('# END SUBSECTION')
     if ret_test:
         return ret_test
 
-    print('# BEGIN SUBSECTION: ament test_results')
+    print('# BEGIN SUBSECTION: colcon test-result')
     # Collect the test results
     ret_test_results = job.run(
-        ['"%s"' % job.python, '-u', '"%s"' % ament_py, 'test_results', '"%s"' % args.buildspace],
+        [args.colcon_script, 'test-result', '--build-base', '"%s"' % args.buildspace, '--all'],
```

Comment:
It is pretty inconvenient to have to open the full console log just to see the summary of the test results. Could we either not pass `--all` or print the summary after the list of test result files?
Ideally I'd like to have both (not passing all and displaying summary after rather than before) as this provides us the same behavior as ament that I find more convenient

Reply:
https://github.com/colcon/colcon-test-result/pull/2

Reply:
Moving the summary below the individual results sounds good to me.

Imo the list of all results is still valuable. There is no other easy way to see which tests ran otherwise. Therefore I would prefer keeping `--all`.

If you think moving the summary below isn't sufficient for readability we could consider to print the test results once with `--all` (for the complete list) and once without (for a brief list of results with errors/failures).

Reply:
I guess it's pretty subjective. I find it more valuable to have a quick summary of what failed directly at the end of the job and to go to the Jenkins test interface if I want the full detail of tests (passing or failing).
But I guess your point is that the jenkins interface already provides an easy way to see failed tests separated from the ran tests? that makes the segregation in the console output less useful/redundant ?

I would be in favor of running the command twice then as it wouldn't impact CI time much and provide easily redable information as (as a developer) I care more often about what failed than what has been ran.

Reply:
Ok: 4030b517f62c4f49bb358fc5dca9591509bbf612